File: benchmark/comparisons/script/scripts/compute_pi.py
# compute_pi.py — Leibniz series approximation of pi.
# Reads `count` (iterations) from sys.argv[1].
# Prints self-measured elapsed ns as the last stdout line.

import sys
import time


def compute_pi(iterations):
    pi = 0.0
    i = 0
    while i < iterations:
        term = 1.0 / (2 * i + 1)
        even = i - (i // 2) * 2
        if even == 0:
            pi += term
        if even == 1:
            pi -= term
        i += 1
    return 4 * pi


count = int(sys.argv[1])

t0 = time.perf_counter_ns()
result = compute_pi(count)
elapsed = time.perf_counter_ns() - t0

print(elapsed)

File: benchmark/comparisons/script/scripts/compute_pi.lua
-- compute_pi.lua -- Leibniz series approximation of pi.
-- Reads count from arg[1]; prints self-measured elapsed ns as last stdout line.
--
-- Lua 5.4 has no built-in nanosecond clock; os.clock() returns CPU seconds
-- with platform-dependent precision (microseconds on Linux). For our
-- pure-CPU workloads CPU time tracks wall time closely.

local function compute_pi(iterations)
    local pi = 0.0
    local i = 0
    while i < iterations do
        local term = 1.0 / (2 * i + 1)
        local even = i - (i // 2) * 2
        if even == 0 then
            pi = pi + term
        end
        if even == 1 then
            pi = pi - term
        end
        i = i + 1
    end
    return 4 * pi
end

local count = tonumber(arg[1])

local t0 = os.clock()
local result = compute_pi(count)
local elapsed_ns = (os.clock() - t0) * 1e9

print(math.floor(elapsed_ns))
